File: C++/Programmers/lv_0/babbling.cpp
#include <string>
#include <vector>

using namespace std;

const string kSpeakableWords[] = { "aya", "ye", "woo", "ma" };

bool IsSpeakable(string word)
{
    int wLen = word.size();
    
    while (true)
    {
        bool foundAtLeastOneWordMatched = false;
        int compCnt = sizeof(kSpeakableWords) / sizeof(kSpeakableWords[0]);
        for (int swi = 0; swi < compCnt; ++swi)
        {
            int swLen = kSpeakableWords[swi].size();
            if (wLen < swLen)
                continue;
            
            bool matched = true;
            for (int wi = 0; wi < swLen; ++wi)
            {
                if (word[wi] != kSpeakableWords[swi][wi])
                {
                    matched = false;
                    break;
                }
            }
            
            if (matched == false)
                continue;
            
            foundAtLeastOneWordMatched = true;
            word.erase(0, swLen);
        }
        
        if (foundAtLeastOneWordMatched == false)
            break;
    }
    
    return word.size() == 0;
}

int solution(vector<string> babbling) {
    int answer = 0;
    int size = babbling.size();
    for (int i = 0; i < size; i++)
    {
        if (IsSpeakable(babbling[i]))
            ++answer;
    }
    return answer;
}

File: C#/Programmers/lv_0/babbling.cs
using System;

public class Solution {
    private static readonly string[] kSpeakableWords = new string[] { "aya", "ye", "woo", "ma" };
    
    public int solution(string[] babbling) {
        int answer = 0;
        for (int i = 0; i < babbling.Length; i++)
        {
            if (IsSpeakable(babbling[i]))
                ++answer;
        }
        return answer;
    }

    private bool IsSpeakable(string word)
    {
        int offset = 0;
        while (offset < word.Length)
        {
            int speakableWordSize = SizeOfFirstSpeakableWordIn(word.Substring(offset));
            if (speakableWordSize == -1)
                return false;

            offset += speakableWordSize;
        }
        return true;
    }

    private int SizeOfFirstSpeakableWordIn(string word)
    {
        if (string.IsNullOrEmpty(word))
            return -1;

        int compCnt = kSpeakableWords.Length;
        for (int swi = 0; swi < compCnt; ++swi)
        {
            if (word.IndexOf(kSpeakableWords[swi]) == 0)
                return kSpeakableWords[swi].Length;
        }
        return -1;
    }
}
